pkg/cache/v3/delta.go

```diff
@@ -30,25 +30,39 @@ type resourceContainer struct {
 
 func createDeltaResponse(ctx context.Context, req *DeltaRequest, state stream.StreamState, resources resourceContainer) *RawDeltaResponse {
 	// variables to build our response with
-	nextVersionMap := make(map[string]string)
-	filtered := make([]types.Resource, 0, len(resources.resourceMap))
-	toRemove := make([]string, 0)
+	var nextVersionMap map[string]string
+	var filtered []types.Resource
+	var toRemove []string
 
 	// If we are handling a wildcard request, we want to respond with all resources
 	switch {
 	case state.IsWildcard():
+		if len(state.GetResourceVersions()) == 0 {
+			filtered = make([]types.Resource, 0, len(resources.resourceMap))
+		}
+		nextVersionMap = make(map[string]string, len(resources.resourceMap))
 		for name, r := range resources.resourceMap {
 			// Since we've already precomputed the version hashes of the new snapshot,
 			// we can just set it here to be used for comparison later
 			version := resources.versionMap[name]
 			nextVersionMap[name] = version
 			prevVersion, found := state.GetResourceVersions()[name]
-			if !found || (prevVersion != nextVersionMap[name]) {
+			if !found || (prevVersion != version) {
 				filtered = append(filtered, r)
 			}
 		}
+
+		// Compute resources for removal
+		for name, prevVersion := range state.GetResourceVersions() {
+			// The prevVersion != "" check is in place to make sure we are only sending an update to the client once right after it is removed.
```

Comment:
I updated the comments to match, and also moved those to non-wildcard only
In the wildcard case the version does not matter as if the object has been removed in a previous iteration it would no longer be present in the version map